### ISSUE TITLE
Mypy fails to typecheck function with two Callable parameters where the output of one is the input of the other

### ISSUE THREAD

**Issue by 8BitMate** [python/mypy#15115]
**Bug Report**

I made a decorator function that takes in a function which checks the output of the decorated function, and I was able to type this using `ParamSpec` `TypeSpec` and `Callable`. This type checks without issue. I did however run into a problem when I created a wrapper function for this decorator which takes both the function and the output value checker function as input parameters and returns the wrapped function. When I run mypy on it it gives some unexpected error messages. In the simplified example below the decorator function is called `foo` and the wrapper function is called `bar`.

**To Reproduce**

```python
from typing import Callable, TypeVar, ParamSpec

P = ParamSpec("P")
R = TypeVar("R")


def foo(return_value_checker: Callable[[R], bool]) -> Callable[[Callable[P, R]], Callable[P, R]]:
    return lambda fun: fun


def bar(fun: Callable[P, R], return_value_checker: Callable[[R], bool]) -> Callable[P, R]:
    return foo(return_value_checker)(fun)
```

**Expected Behavior**

I expect this to type check without errors.

**Actual Behavior**

Output when running mypy:

```
generic_paramspec.py:12: error: Incompatible return value type (got "Callable[P, P]", expected "Callable[P, R]")  [return-value]
generic_paramspec.py:12: error: Argument 1 has incompatible type "Callable[P, R]"; expected "Callable[P, P]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.2.0
- Python version used: 3.10.9

**Comment by ilevkivskyi** [comment]
This works on master.

**Comment by hauntsaninja** [comment]
@8BitMate is the project you were running into this on open source?

**Comment by 8BitMate** [comment]
@hauntsaninja No it is a closed source project at work.
